### DISCUSSION TITLE
Thread safety of SQLAlchemy sessions for read-only

### DISCUSSION THREAD

**Discussion by vmandke** [sqlalchemy/sqlalchemy#10362] [Usage Questions]
From docs, and SO discussions it seems that the Session object is not threadsafe. However, it uses the QueuePool by default. If no flushes or commits are sent to DBAPI, essentially the Session object is thread-safe, right? For read-only purposes?

**Comment by CaselIT** [comment]
Hi,

No, it's not since if you run two select at the same time you can still have issues. You should still use different sessions even if read only

**Reply by vmandke** [reply]
won't the issue only surface if the cache_key of both the selects is the same?

**Reply by vmandke** [reply]
I'm looking at the query code, please let me know if I am missing something here. (trying to understand this more, as I haven't run into data issues and I am using the same Session across threads :) )

**Reply by CaselIT** [reply]
> won't the issue only surface if the cache_key of both the selects is the same?

that's not the main issue here, the problem is that you have a single connection to the db and you are trying to use it from multiple thread. In general that's does not work, since you could have cases where the db is sending result to one thread and you send a new query from another.

Some dbapi use locks to avoid concurrent access, but accessing the same connection from multiple threads is undefined behaviour.

> trying to understand this more, as I haven't run into data issues and I am using the same Session across threads :)

do as you please, but don't complain if you get errors or strange data back... If you use the session in read only mode I see no reason at all why you would want share a single session, just create one when you need it in the thread.
In the best case (when the dbapi uses locks) you get no benefit at all from using threading

**Comment by zzzeek** [comment]
the ORM Session keeps track of all ORM objects loaded / persisted using an [identity map](https://docs.sqlalchemy.org/en/20/glossary.html#term-identity-map).  This is a stateful object and is not threadsafe.  The `Session` also has the notion of a `SessionTransaction` that's associated with it at all times.   The management of this `SessionTransaction` is also internal state for the `Session` and is not threadsafe.

In short, **the ORM Session is not threadsafe in any way, shape, or form, and if your program "works", that is an illusion, it will break in silent, difficult to debug ways if you are using a single Session in more than one thread**.
